diff --git a/rust/platform/triple_mappings.bzl b/rust/platform/triple_mappings.bzl
index 0614eac..04e9502 100644
--- a/rust/platform/triple_mappings.bzl
+++ b/rust/platform/triple_mappings.bzl
@@ -24,6 +24,7 @@ SUPPORTED_T2_PLATFORM_TRIPLES = [
     "aarch64-unknown-linux-gnu",
     "arm-unknown-linux-gnueabi",
     "armv7-unknown-linux-gnueabi",
+    "armv7-linux-androideabi",
     "i686-linux-android",
     "i686-unknown-freebsd",
     "powerpc-unknown-linux-gnu",
@@ -63,6 +64,7 @@ _CPU_ARCH_TO_BUILTIN_PLAT_SUFFIX = {
 # Systems that map to a "@platforms//os entry
 _SYSTEM_TO_BUILTIN_SYS_SUFFIX = {
     "android": "android",
+    "androideabi": "android",
     "bitrig": None,
     "darwin": "osx",
     "dragonfly": None,
@@ -82,6 +84,7 @@ _SYSTEM_TO_BUILTIN_SYS_SUFFIX = {
 
 _SYSTEM_TO_BINARY_EXT = {
     "android": "",
+    "androideabi": "",
     "darwin": "",
     "eabi": "",
     "emscripten": ".js",
@@ -98,6 +101,7 @@ _SYSTEM_TO_BINARY_EXT = {
 
 _SYSTEM_TO_STATICLIB_EXT = {
     "android": ".a",
+    "androideabi": ".a",
     "darwin": ".a",
     "eabi": ".a",
     "emscripten": ".js",
@@ -111,6 +115,7 @@ _SYSTEM_TO_STATICLIB_EXT = {
 
 _SYSTEM_TO_DYLIB_EXT = {
     "android": ".so",
+    "androideabi": ".so",
     "darwin": ".dylib",
     "eabi": ".so",
     "emscripten": ".js",
@@ -127,6 +132,7 @@ _SYSTEM_TO_STDLIB_LINKFLAGS = {
     # NOTE: Rust stdlib `build.rs` treats android as a subset of linux, rust rules treat android
     # as its own system.
     "android": ["-ldl", "-llog", "-lgcc"],
+    "androideabi": ["-ldl", "-lgcc"],
     "bitrig": [],
     # TODO(gregbowyer): If rust stdlib is compiled for cloudabi with the backtrace feature it
     # includes `-lunwind` but this might not actually be required.
