diff --git a/rust/private/rust.bzl b/rust/private/rust.bzl
index def840e..8f9ff00 100644
--- a/rust/private/rust.bzl
+++ b/rust/private/rust.bzl
@@ -556,6 +556,7 @@ _common_attrs = {
             is the root of the crate to be passed to rustc to build this crate.
         """),
         allow_files = [".rs"],
+        flags = ["DIRECT_COMPILE_TIME_INPUT"],
     ),
     "stamp": attr.int(
         doc = dedent("""\
